sel/src/Sel/SecretKey/EncryptedStream.hs

```diff
@@ -0,0 +1,542 @@
+import Foreign.C (CChar, CSize, CUChar, CULLong)
+import GHC.IO.Handle.Text (memcpy)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+import Data.Text (Text)
+import LibSodium.Bindings.SecretStream
+  ( CryptoSecretStreamXChaCha20Poly1305State
+  , cryptoSecretStreamXChaCha20Poly1305ABytes
+  , cryptoSecretStreamXChaCha20Poly1305HeaderBytes
+  , cryptoSecretStreamXChaCha20Poly1305InitPull
+  , cryptoSecretStreamXChaCha20Poly1305InitPush
+  , cryptoSecretStreamXChaCha20Poly1305KeyBytes
+  , cryptoSecretStreamXChaCha20Poly1305KeyGen
+  , cryptoSecretStreamXChaCha20Poly1305Pull
+  , cryptoSecretStreamXChaCha20Poly1305Push
+  , cryptoSecretStreamXChaCha20Poly1305Rekey
+  , cryptoSecretStreamXChaCha20Poly1305StateBytes
+  , cryptoSecretStreamXChaCha20Poly1305TagFinal
+  , cryptoSecretStreamXChaCha20Poly1305TagMessage
+  , cryptoSecretStreamXChaCha20Poly1305TagPush
+  , cryptoSecretStreamXChaCha20Poly1305TagRekey
+  )
+import Sel.Internal
+
+-- | The 'SecretKey' is used to encrypt the stream.
+--
+-- @since 0.0.1.0
+newtype SecretKey = SecretKey (ForeignPtr CUChar)
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+      -- > display secretKey == "[REDACTED]"
+    )
+    via (OpaqueInstance "[REDACTED]" SecretKey)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq SecretKey where
+  (SecretKey hk1) == (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord SecretKey where
+  compare (SecretKey hk1) (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
+
+-- | > show secretKey == "[REDACTED]"
+--
+-- @since 0.0.1.0
+instance Show SecretKey where
+  show _ = "[REDACTED]"
+
+-- | Generate a new random secret key.
+--
+-- @since 0.0.1.0
+newSecretKey :: IO SecretKey
+newSecretKey = do
+  fPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretStreamXChaCha20Poly1305KeyBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    cryptoSecretStreamXChaCha20Poly1305KeyGen ptr
+  pure $ SecretKey fPtr
+
+-- == CipherText ==
+
+-- | A 'CipherText' holds an encrypted message and its own length.
+--
+-- @since 0.0.1.0
+data CipherText
+  = CipherText
+      !(ForeignPtr CUChar)
+      -- ^ Content of the ciphertext
+      !CSize
+      -- ^ Length of the ciphertext
+
+-- | Convert a binary 'StrictByteString' from the outside (filesystem, network) to a 'CipherText'.
+--
+-- If the message is not strictly longer than the size of the authentication tag,
+-- the function will fail.
+--
+-- @since 0.0.1.0
+cipherTextFromByteString :: StrictByteString -> Maybe CipherText
+cipherTextFromByteString bytestring =
+  let (foreignPtr, bsLength) = BS.toForeignPtr0 bytestring
+   in if bsLength > fromIntegral @CSize @Int cryptoSecretStreamXChaCha20Poly1305ABytes
+        then Just $ CipherText (Foreign.castForeignPtr @Word8 @CUChar foreignPtr) (fromIntegral @Int @CSize bsLength)
+        else Nothing
```

Comment:
I'm not sure this is completely safe: we copy a pointer to some foreign data, but not the data itself, which means that when the `ByteString` drops out of scope, our pointer now dangles. An 'owning copy' would be much better.

Reply:
Something like this?

```haskell
      cipherForeignPtr <- Foreign.mallocPlainForeignPtrBytes (BS.length bytestring)
      Foreign.withForeignPtr cipherForeignPtr $ \cipherPtr ->
        BS.unsafeUseAsCStringLen bytestring $ \(cString, cStringLen) -> do
          memcpy (Foreign.castPtr cipherPtr) cString (fromIntegral cStringLen)
          pure $
            Just $
              CipherText
                cipherForeignPtr
                (fromIntegral @Int @CSize cStringLen)
```

Reply:
Does `mallocPlainForeignPtrBytes` give us a finalizer?